### PR TITLE
Implement DownloadStream

### PR DESCRIPTION
**Pull request by Jazz86**

### CODE REVIEW

File: src/Header/DownloadStream.h [hunk 1]
```diff
@@ -25,7 +25,9 @@ class DownloadStream {
   // filename include file path
   std::string filename;
   // default chunk size
-  unsigned long chuck_size;
+  unsigned long chunk_size;
+  // True indicate that it has the read access of the file, otherwise.
+  bool reading = false;
```
Comment (younger15) [inline src/Header/DownloadStream.h]: do we initialize variable in class declaration?

Comment (Jazz86) [reply src/Header/DownloadStream.h]: Not sure the google style mention about this, it has been fixed, after all.

**Review by younger15 — APPROVED** [review]

**Review by LinRenHong — APPROVED** [review]

File: src/Implement/DownloadStream.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+int DownloadStream::ReadChunk(unsigned long &pos, std::string *const binary) {
+  if (!reading) {
+    return 0;
+  }
+
+  fd.seekg(pos * chunk_size, std::ios::beg);
+  char buffer[chunk_size + 1] = "";
+  fd.read(buffer, chunk_size);
+  *binary = buffer;
+  if (fd.eof()) {
```
Comment (john0312) [inline src/Implement/DownloadStream.cpp:26]: You might need to check the fail bit and bad bit as well.

File: src/Implement/DownloadStream.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+int DownloadStream::ReadChunk(unsigned long &pos, std::string *const binary) {
+  if (!reading) {
+    return 0;
+  }
+
+  fd.seekg(pos * chunk_size, std::ios::beg);
+  char buffer[chunk_size + 1] = "";
+  fd.read(buffer, chunk_size);
+  *binary = buffer;
```
Comment (john0312) [inline src/Implement/DownloadStream.cpp:25]: string(const char*) construct the string with NULL terminated strings. This might not work if the file contains '\0'.

File: src/Implement/UnitTestDownloadStream.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <gtest/gtest.h>
+
+#include "DownloadStream.h"
+
+// test basic usage
+TEST(DOWNLOADSTREAM_TEST, Basic) {
+  // create temporary test file
+  std::fstream newfile("test", std::fstream::out);
+  newfile << "123456";
```
Comment (john0312) [inline src/Implement/UnitTestDownloadStream.cpp]: Please make the test string a constant. i.e. `constexpr char kTestFileContent[]`, this is so that we've the flexibility of changing the test string.
Also, consider adding characters such as '\0' and '\n' into the test string, this is a common corner case for many string related code.

File: src/Implement/UnitTestDownloadStream.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  newfile.close();
+
+  DownloadStream down("test", 1);
+  unsigned long pos = 0;
+  std::string binary;
+  int result;
+
+  // check ReadChunk
+  for (pos = 0; pos < 6; pos++) {
```
Comment (john0312) [inline src/Implement/UnitTestDownloadStream.cpp]: Consider using `sizeof(kTestFileContent)` instead of 6, this way, we change easily change the content of the test string without changing the test.

File: src/Header/DownloadStream.h [hunk 6]
```diff
@@ -25,7 +26,9 @@ class DownloadStream {
   // filename include file path
   std::string filename;
   // default chunk size
-  unsigned long chuck_size;
+  unsigned long chunk_size;
+  // True indicate that it has the read access of the file, otherwise.
+  bool reading;
```
Comment (john0312) [inline src/Header/DownloadStream.h]: [Optional] std::optional<std::fstream> can be used here instead for one lesser variable.

File: src/Implement/DownloadStream.cpp (hunk 3 above)
Comment (Jazz86) [reply src/Implement/DownloadStream.cpp:25]: Assuming the file is plaintext, there won't be '\0' in the middle of file content.

File: src/Header/DownloadStream.h (hunk 6 above)
Comment (Jazz86) [reply src/Header/DownloadStream.h]: Removed. Using fd.fail() to check badbit and failbit.

File: src/Implement/DownloadStream.cpp (hunk 2 above)
Comment (Jazz86) [reply src/Implement/DownloadStream.cpp:26]: Using fd.fail() to check badbit and failbit at upper line.

File: src/Implement/UnitTestDownloadStream.cpp (hunk 5 above)
Comment (Jazz86) [reply src/Implement/UnitTestDownloadStream.cpp]: done.

File: src/Implement/UnitTestDownloadStream.cpp (hunk 4 above)
Comment (Jazz86) [reply src/Implement/UnitTestDownloadStream.cpp]: done. Assuming the file is plaintext, there won't be '\0' in the middle of file content.
